pcdsdevices/gon.py

```diff
@@ -236,23 +255,348 @@ class Kappa(BaseInterface, Device):
+        k_phi = -k_phi * 180 / np.pi
+        return k_eta, k_kap, k_phi
+
+    def mv_e_eta(self, value):
+        """
+        Change the e_eta position to value, keeping e_chi and e_phi the same.
+
+        Parameters
+        ----------
+        value : number
+            Position value to set e_eta to.
+        """
+        try:
+            k_eta, k_kap, k_phi = self.e_to_k(e_eta=value)
+            if self.check_motor_step(k_eta, k_kap, k_phi):
+                logger.info("Starting now moving things!!!")
+                self.eta.mv(k_eta)
+                self.kappa.mv(k_kap)
+                self.phi.mv(k_phi)
+                self.wait()
+        except KeyboardInterrupt:
+            logger.info("Motion interrupted by ctrl+c")
+            self.stop()
+
+    def mv_e_chi(self, value):
+        """
+        Change the e_chi position to value, keeping e_eta and e_phi the same.
+
+        Parameters
+        ----------
+        value : number
+            Position value to set e_chi to.
+        """
+        try:
+            if value == 0:
+                value = 10e-9
+            k_eta, k_kap, k_phi = self.e_to_k(e_chi=value)
+            if self.check_motor_step(k_eta, k_kap, k_phi):
+                logger.info("Starting now moving things!!!")
+                self.eta.mv(k_eta)
+                self.kappa.mv(k_kap)
+                self.phi.mv(k_phi)
+                self.wait()
+        except KeyboardInterrupt:
+            logger.info("Motion interrupted by ctrl+c")
+            self.stop()
+
+    def mv_e_phi(self, value):
+        """
+        Change the e_phi position to value, keeping e_eta and e_phi the same.
+
+        Parameters
+        ----------
+        value : number
+            Position value to set the e_phi to.
+        """
+        try:
+            k_eta, k_kap, k_phi = self.e_to_k(e_phi=value)
+            if self.check_motor_step(k_eta, k_kap, k_phi):
+                logger.info("Starting now moving things!!!")
+                self.eta.mv(k_eta)
+                self.kappa.mv(k_kap)
+                self.phi.mv(k_phi)
+        except KeyboardInterrupt:
+            logger.info("Motion interrupted by ctrl+c")
+            self.stop()
+
+    @pseudo_position_argument
+    def forward(self, pseudo_pos):
+        """
+        Calculate a RealPosition from a given PseudoPosition.
+
+        Parameters
+        ----------
+        pseudo_pos : PseudoPosition
+            The pseudo position input.
+        Returns
+        -------
+        real_position : RealPosition
+            The real position output.
+        """
+        pseudo_pos = self.PseudoPosition(*pseudo_pos)
+        eta, kappa, phi = self.e_to_k(pseudo_pos.e_eta, pseudo_pos.e_chi,
+                                      pseudo_pos.e_phi)
+        return self.RealPosition(eta=eta, kappa=kappa, phi=phi,
+                                 x=self.x.position, y=self.y.position,
+                                 z=self.z.position)
```

Comment:
How do we go about this situation where I have more real motors than pseudo motors and don't need to map all o them?
Do I set them to `None`? 

```
return self.RealPosition(eta=eta, kappa=kappa, phi=phi,
                                 x=None, y=None,
                                 z=None)
```
or do i just take their `position`?

Reply:
Stepping back for a second, the intent `PseudoPositioner` is that there's always a mapping of pseudo<->real positioners. If they are not needed, why are they in the positioner in the first place?

There is a mechanism to exclude these positioners from the real positioner list, but I believe it does not work (since I wrote the thing in 2015, I only just heard of one person trying and failing to use it this year). So, just an idea:

1. Create a `KappaSample` `Device` with the XYZ stages (if it's not a positioner, it won't get picked up by the `PseudoPositioner` mechanism)
2. On `Kappa` init, set `self.x = self.sample_stage.x` for user convenience

Reply:
OK, I'll try this - thanks @klauer
(in the old code they have x, y, and z as optional (`samx`, `samy`, `samz`)) - this might have solved this problem

Reply:
I thought I understood what I need to do but I got stuck a bit. :( 

Could I just use the `XYZStage` class instead of creating a `KappaSample`? 
https://github.com/pcdshub/pcdsdevices/blob/fa9ddd03634bd5d80bddb04dd83ff8967cd26ccd/pcdsdevices/gon.py#L154-L183

Is `sample_stage` a new component I have to add besides the x, y, and z?
```
sample_stage = Cpt(XYZStage, ...)
```
and have the user give a prefix for this guy now at init?

Reply:
Looks like it's already there - good catch! Definitely we should reuse devices where possible.

`FormattedComponent` is a bit faulty here, I think you're finding - you have to know at `__init__` time what the prefixes are, but you have no control over `XYZStage.__init__` in your `Kappa` class.

There may be other/better ways; my first thought is that the following minimal subclass might work:
```python
class KappaXYZStage(XYZStage):
   def __init__(self, *args, parent, **kwargs):
       self._prefix_x = parent._sample_x_prefix  # set `_sample_x_prefix` in `Kappa.__init__` before calling `super().__init__`
       # repeat for y and z prefix
       super().__init__(*args, parent=parent, **kwargs)
```

Reply:
Ok thank you, this `KappaXYZStage` seems to work but I can't figure out how to set the `x` to `sample_stage.x` - i'm probably doing something very stupid here...

"On Kappa init, set self.x = self.sample_stage.x for user convenience" - I don't know of `sample_stage` in the `Kappa` `__init__` so can't set it.... 

I have this class:
```
class KappaXYZStage(XYZStage):
    def __init__(self, *args, parent, **kwargs):
        self._prefix_x = parent._prefix_x
        self._prefix_y = parent._prefix_y
        self._prefix_z = parent._prefix_z
        super().__init__(*args, parent=parent, prefix_x=self._prefix_x,
                         prefix_y=self._prefix_y, prefix_z=self._prefix_z,
                         **kwargs)
```
and then in `Kappa`:
```
    x = FCpt(IMS, '{self._prefix_x}', kind='normal')
    y = FCpt(IMS, '{self._prefix_y}', kind='normal')
    z = FCpt(IMS, '{self._prefix_z}', kind='normal')

    sample_stage = Cpt(KappaXYZStage, name='xyz_stage', kind='omitted')


.....

    def __init__(self, *, name, prefix_x, prefix_y, prefix_z,
                 prefix_eta, prefix_kappa, prefix_phi, eta_max_step=2,
                 kappa_max_step=2, phi_max_step=2, kappa_ang=50, **kwargs):
        self._prefix_x = prefix_x
        self._prefix_y = prefix_y
        self._prefix_z = prefix_z
        self._prefix_eta = prefix_eta
        self._prefix_kappa = prefix_kappa
        self._prefix_phi = prefix_phi
        self.eta_max_step = eta_max_step
        self.kappa_max_step = kappa_max_step
        self.phi_max_step = phi_max_step
        self.kappa_ang = kappa_ang
        # self.x = ..
        # self.y = ...
        # self.z = ...
        super().__init__('', name=name, **kwargs)
```

Reply:
The key is that `self.x` - the instantiated component, or the `IMS` device instance - will only be available after `super().__init__`. Moving those lines down should do it: `self.x = self.sample_stage.x`

Reply:
Ohh ok, I tried that and I was getting an error and did not think of trying it again: 

```
--> 340         self.x = self.sample_stage.x
    341         self.y = self.sample_stage.y
    342         self.z = self.sample_stage.z

~/anaconda3/envs/bluesky/lib/python3.8/site-packages/ophyd/device.py in __set__(self, instance, owner)
    226 
    227     def __set__(self, instance, owner):
--> 228         raise RuntimeError('Do not use setattr with components; use '
    229                            'cpt.put(value)')
    230 

RuntimeError: Do not use setattr with components; use cpt.put(value)
```

Reply:
That's correct, but that error is coming from `x`, `y`, and `z` till being components of `Kappa` (it thinks you're trying to do `self.x.put(self.sample_stage.x)`.

Reply:
Ok thanks a lot Ken for spoon-feeding me on this Issue - I thought I'd get it done sooner because they need this on Friday :( but I'm still trying to figure it out

Reply:
No worries - and if you need, let me know how I can help you push it over the finish line. I'm sure we can get it working by Friday.

Reply:
Ok thank you, i'll try to see if I can make it work until tomorrow, if not I would love some help with it - thanks a lot. 

For now if I do:
```
  super().__init__(name=name, **kwargs)
        Kappa.x = self.sample_stage.x
        Kappa.y = self.sample_stage.y
        Kappa.z = self.sample_stage.z
```
```
In [4]: from pcdsdevices.gon import Kappa

In [5]: k = Kappa(name='name', prefix_x='x', prefix_y='y', prefix_z='z', prefix_eta='eta', prefix_kappa='kappa', prefix_phi='phi')

In [6]: k
Out[6]: 

name
----
position: ERROR

sample_stage_x
--------------
  position: ERROR
  user_setpoint: None
  description: None
  dial_position: None

sample_stage_y
--------------
  position: ERROR
  user_setpoint: None
  description: None
  dial_position: None

sample_stage_z
--------------
  position: ERROR
  user_setpoint: None
  description: None
  dial_position: None

eta
---
  position: ERROR
  user_setpoint: None
  description: None
  dial_position: None

kappa
-----
  position: ERROR
  user_setpoint: None
  description: None
  dial_position: None

phi
---
  position: ERROR
  user_setpoint: None
  description: None
  dial_position: None

e_eta
-----
  position: ERROR
  notepad_setpoint: 0.0
  notepad_readback: 0.0

e_chi
-----
  position: ERROR
  notepad_setpoint: 0.0
  notepad_readback: 0.0

e_phi
-----
  position: ERROR
  notepad_setpoint: 0.0
  notepad_readback: 0.0
```
it seems to set the `x`, `y` and `z` fine (i think) but it still sees them as `positioners` this way

I'll keep trying till tomorrow and ask for help tomorrow ;)

Reply:
(Resolved over chat)